server/raft_transport.go

```diff
@@ -70,11 +83,50 @@ func newRPCTransport(gossip *gossip.Gossip, rpcServer *rpc.Server, rpcContext *r
 			t.RaftMessage, &multiraft.RaftMessageRequest{}); err != nil {
 			return nil, err
 		}
+
+		if err := t.rpcServer.RegisterAsync(multiPackageMessageName, false, /*not public*/
+			t.MultiPackageMessage, &multiraft.MultiPackageMessageRequest{}); err != nil {
+			return nil, err
+		}
 	}
 
 	return t, nil
 }
 
+// MultiPackageMessage merges incoming multi-package request into the real request,
+// and proxies to the listening server interface.
+func (t *rpcTransport) MultiPackageMessage(args proto.Message, callback func(proto.Message, error)) {
+	req := args.(*multiraft.MultiPackageMessageRequest)
+
+	t.multiPackageMessageLock.Lock()
+	_, ok := t.multiPackageMessage[req.GroupID]
+	if !ok {
+		t.multiPackageMessage[req.GroupID] = &[]byte{}
+	}
+	data := t.multiPackageMessage[req.GroupID]
+	*data = append(*data, req.RawBytes...)
```

Comment:
@bdarnell If ```MultiPackageMessage``` receive request's PackageID is 0, I drop the old data if exist and restart to cache the new data until all packages received.
But the problem is that if leader never send a snapshot message after failing, this data will be always keep in memory.